File: core/src/main/java/com/vzome/core/edits/TransformSelection.java

package com.vzome.core.edits;

import java.util.ArrayList;
import java.util.List;

import com.vzome.core.construction.Construction;
import com.vzome.core.construction.Transformation;
import com.vzome.core.editor.api.ChangeManifestations;
import com.vzome.core.editor.api.EditorModel;
import com.vzome.core.model.Manifestation;

public class TransformSelection extends ChangeManifestations
{
    protected final Transformation transform;
    
    public TransformSelection( EditorModel editor, Transformation transform )
    {
        super( editor );
        this .transform = transform;
    }
    
    @Override
    public void perform()
    {
        List<Manifestation> inputs = new ArrayList<>();
        for (Manifestation man : mSelection) {
            unselect( man );
            inputs .add( man );
        }
        
        redo();  // get the unselects out of the way, if anything needs to be re-selected
        // now apply it to the input objects
        for (Manifestation m : inputs) {
            if ( ! m .isRendered() )
                continue;
            Construction c = m .getFirstConstruction();
            Construction result = transform .transform( c );
            select( manifestConstruction( result ), true );
        }
        redo();
    }
    
    @Override
    protected String getXmlElementName()
    {
        return "TransformSelection";
    }
}

File: online/src/worker/legacy/from-java/com/vzome/core/edits/TransformSelection.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { Construction } from "../construction/Construction.js";
import { Transformation } from "../construction/Transformation.js";
import { ChangeManifestations } from "../editor/api/ChangeManifestations.js";
import { EditorModel } from "../editor/api/EditorModel.js";
import { Manifestation } from "../model/Manifestation.js";

export class TransformSelection extends ChangeManifestations {
    transform: Transformation;

    public constructor(editor: EditorModel, transform: Transformation) {
        super(editor);
        if (this.transform === undefined) { this.transform = null; }
        this.transform = transform;
    }

    /**
     * 
     */
    public perform() {
        const inputs: java.util.List<Manifestation> = <any>(new java.util.ArrayList<any>());
        for(let index=this.mSelection.iterator();index.hasNext();) {
            let man = index.next();
            {
                this.unselect$com_vzome_core_model_Manifestation(man);
                inputs.add(man);
            }
        }
        this.redo();
        for(let index=inputs.iterator();index.hasNext();) {
            let m = index.next();
            {
                if (!m.isRendered())continue;
                const c: Construction = m.getFirstConstruction();
                const result: Construction = this.transform.transform$com_vzome_core_construction_Construction(c);
                this.select$com_vzome_core_model_Manifestation$boolean(this.manifestConstruction(result), true);
            }
        }
        this.redo();
    }

    /**
     * 
     * @return {string}
     */
    getXmlElementName(): string {
        return "TransformSelection";
    }
}
TransformSelection["__class"] = "com.vzome.core.edits.TransformSelection";
